gprofiler/profilers/java.py

```diff
@@ -257,14 +257,45 @@ def is_java_basename(process: Process) -> bool:
 _JAVA_VERSION_CACHE_MAX = 1024
 
 
+def _get_process_ns_java_path(process: Process) -> Optional[str]:
+    """
+    Look up path to java executable installed together with this process' libjvm.
+    """
+    java_path: Optional[str] = None
+    nspid = get_process_nspid(process.pid)
+    if is_java_basename(process):
+        java_path = f"/proc/{nspid}/exe"  # it's a symlink and will be resolveable under process' mnt ns
+    else:
+        libjvm_path: Optional[str] = None
+        for m in process.memory_maps():
+            if re.match(DETECTED_JAVA_PROCESSES_REGEX, m.path):
+                libjvm_path = m.path
+                break
+        if libjvm_path is None:
+            return None
+        libjvm_dir = os.path.dirname(libjvm_path)
+        # support two java layouts - it's either lib/server/../../bin/java or lib/{arch}/server/../../../bin/java:
+        java_candidate_paths = [
+            Path(libjvm_dir, "../../bin/java").resolve(),
+            Path(libjvm_dir, "../../../bin/java").resolve(),
+        ]
+        for p in java_candidate_paths:
+            proc_relative_path = Path(f"/proc/{process.pid}/root", p.relative_to("/"))
```

Comment:
You need to use `resolve_proc_root_links` here, otherwise this naive join of paths might escape symlinks. Please read that function's docs to understand how to use, and use it here.

Reply:
That function call is not needed here - we're basing search on libjvm path taken from `process.memory_maps()`, which is normalized by system.